### ISSUE TITLE
Running models on bulk nanopore data?

### ISSUE THREAD

**Issue by Jonbroad15** [gdewael/cpg-transformer#1]
Hi, I was wondering if I could run this program on bulk nanopore data?

I am trying to impute cpg sites to run a methylation atlas on cell free DNA data to determine cell of origin. Our data gives modification frequencies for each coordinate in the genome, however we need to fill in some coordinates to cover all the sites in the methylation atlas.

Please let me know your recommendations in using cpg-transformer for this task.

**Comment by gdewael** [comment]
Hi Jonathan,

I think CpG Transformer should be applicable to any incomplete (eukaryotic) methylation data source, whether it be nanopore or bisulfite, or bulk or single-cell.
Ideally, you should have a data matrix that looks somewhat like Fig. 1 (left) in our [publication](https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btab746/6413629)
I think in your case, it will be easiest to get data in a .tsv format that looks like described [here](https://github.com/gdewael/cpg-transformer/tree/main/data#from-tsv-files).

If I understand correctly, you are looking to match your _n_ samples each to one or more 'samples/cells' in a methylation atlas (consisting of _m_ cells).
If this is the case: you have two choices: either (1) you only give your _n_ samples as inputs to the models to learn relations between those _n_ samples and impute based on those relations, or (2) you append the methylation atlas as extra 'rows' and give that to the model, in this case, CpG Transformer will additionally already model the relations between your samples and the methylation atlas. (You can disentangle how the relations look like with our interpretation module). If the number of rows is very large (say, more than 100), you will have to adjust some parameters. Prime ones to look at in that case are: `--window` & `--resample_cells(_val)`

Once you obtain the data matrix with one of those two setups, I believe you should be able to run CpG Transformer as normally. I recommend splitting chromosome 5 and 10 as validation and test sets, respectively. You can also try a pre-trained model (we have ones for a variety of biological setups, ideally choose the one closest to yours).

Good luck!

**Comment by Jonbroad15** [comment]
Thanks for the information Gaetan. That really helps a lot. I will try that. The only thing I am worried about is that our methylation calls are not binary. We get methylation frequency (0 <= meth_freq <= 1) for each coordinate. Will cpg-transformer only work on binary methylation calls?

**Comment by gdewael** [comment]
Currently, CpG Transformer only works with methylation matrices with discrete values (either 0, 1 or -1 for unknown).
That is not to say, however, that it is not possible to implement support for continuous methylation frequencies. No promises in regards to the performance though. (In fact, your dataset/usecase can serve as an initial benchmark if you'd like)
If you want, you can open a pull request and I'll start working on support for continuous methylation labels when I can.

**Comment by Jonbroad15** [comment]
Thanks Gaetan. Unfortunately our data has a privacy policy, however I will work on simulating a cfdna dataset to serve as a benchmark for this task.

**Comment by gdewael** [comment]
Hi Jonathan,

Support for continuous methylation frequencies have been added (To encode: see `data/EncodeFromTsv.py`, all the other scripts should run out-of-the-box, but I can't guarantee that transfer learning will give any advantages between the available classification models and the regression models you wish to train). I will now close this issue, if you encounter any other issues, please open a new one. Good luck!

**Comment by Jonbroad15** [comment]
Thanks so much Gaetan, I will give it a try and let you know how it goes!
